File: modules/yomoney-sdk/android/src/main/java/expo/modules/yomoneysdk/YomoneySdkModule.kt
package expo.modules.yomoneysdk


import android.app.Activity
import android.content.Intent
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import ru.yoomoney.sdk.kassa.payments.Checkout.createTokenizationResult
import ru.yoomoney.sdk.kassa.payments.Checkout.createTokenizeIntent
import ru.yoomoney.sdk.kassa.payments.checkoutParameters.PaymentMethodType
import ru.yoomoney.sdk.kassa.payments.checkoutParameters.PaymentParameters
import ru.yoomoney.sdk.kassa.payments.checkoutParameters.SavePaymentMethod
import ru.yoomoney.sdk.kassa.payments.checkoutParameters.Amount
import ru.yoomoney.sdk.kassa.payments.checkoutParameters.TestParameters
import java.math.BigDecimal
import java.util.Currency


class TokenizeOptions : Record {
  @Field val amount: Double = 0.0
  @Field val currency: String = "RUB"
  @Field val title: String = ""
  @Field val subtitle: String = ""
  @Field val clientApplicationKey: String = ""
  @Field val shopId: String = ""
}
const val REQUEST_CODE = 42069

class YomoneySdkModule : Module() {

  private var pendingPromise: Promise? = null

  override fun definition() = ModuleDefinition {
    Name("YomoneySdk")

    OnActivityResult { activity, payload ->
      when(payload.resultCode){
        Activity.RESULT_OK -> handleActivityResult(payload.data)
        Activity.RESULT_CANCELED -> pendingPromise = null
      }
    }

    AsyncFunction("startTokenize") { options: TokenizeOptions, promise: Promise ->

      val currentActivity = appContext.currentActivity
        ?: return@AsyncFunction promise.reject(Exceptions.MissingActivity())

      if (pendingPromise != null) {
        return@AsyncFunction promise.reject("PENDING_PAYMENT",
          "Another payment process is active", null)
      }

      pendingPromise = promise

      val paymentParameters = PaymentParameters(
        amount = Amount(
          BigDecimal.valueOf(options.amount),
          Currency.getInstance(options.currency)
        ),
        title = options.title,
        subtitle = options.subtitle,
        clientApplicationKey = options.clientApplicationKey,
        shopId = options.shopId,
        savePaymentMethod = SavePaymentMethod.ON,
        paymentMethodTypes = setOf(
          PaymentMethodType.BANK_CARD,
          PaymentMethodType.SBP
        ),
      )

      val intent = createTokenizeIntent(currentActivity,
        paymentParameters,
        TestParameters(showLogs = true))

      currentActivity.startActivityForResult(intent,
        REQUEST_CODE)
    }
  }

  private fun handleActivityResult(data: Intent?) {
    val promise = pendingPromise ?: return
    pendingPromise = null
    if(data != null){
      val result = createTokenizationResult(data)
      val response = mapOf(
        "paymentToken" to result.paymentToken,
        "paymentMethod" to result.paymentMethodType
      )
      promise.resolve(response)
    }
  }
}

File: modules/yomoney-sdk/ios/YomoneySdkModule.swift
import ExpoModulesCore
import YooKassaPayments

struct TokenizeOptions : Record {
    @Field var amount: Double = 0.0
    
    @Field var currency: String = "RUB"

    @Field var title: String = ""

    @Field var subtitle: String = ""

    @Field var clientApplicationKey: String = ""
    
    @Field var shopId: String = ""
}

public class YomoneySdkModule: Module, TokenizationModuleOutput {
    
    private var pendingPromise: Promise?
    private var tokenizationVC: UIViewController?
    
    
    public func didFinish(on module: TokenizationModuleInput, with error: YooKassaPaymentsError?) {
        let promise = pendingPromise
        pendingPromise = nil
        
        DispatchQueue.main.async {
          self.tokenizationVC?.dismiss(animated: true)
          self.tokenizationVC = nil
        }
        
        if let error{
            promise?.reject("TOKENIZATION_ERROR", error.localizedDescription)
        } else{
            promise?.resolve(nil)
        }
    }
    
    public func didFinishConfirmation(paymentMethodType: PaymentMethodType) {
        
    }
    
    public func didFailConfirmation(error: YooKassaPaymentsError?) {
        
    }
    
    public func tokenizationModule(_ module:TokenizationModuleInput,
                                   didTokenize token: Tokens,
                                   paymentMethodType: PaymentMethodType) {
        let promise = pendingPromise
        pendingPromise = nil
        
        DispatchQueue.main.async {
            self.tokenizationVC?.dismiss(animated: true){
                self.tokenizationVC = nil
                let response: [String: Any] = [
                    "paymentToken": token.paymentToken,
                    "paymentMethod": paymentMethodType.rawValue
                ]
                promise?.resolve(response)
            }
        }
    }
    
    
//    MARK: - Module defination
    public func definition() -> ModuleDefinition {
        
        Name("YomoneySdk")
        
        AsyncFunction("startTokenize"){ (options: TokenizeOptions, promise: Promise) in
            
            DispatchQueue.main.async {
                guard let currentVC = self.appContext?.utilities?.currentViewController() else {
                    return promise.reject("ERR_MISSING_VC", "No Available ViewController")
                }
                
                guard self.pendingPromise == nil else {
                    return promise.reject("PENDING_PAYMENT",
                      "Another payment process is active")
                }
                
                self.pendingPromise = promise
                
                let amount = Amount(
                    value: Decimal(options.amount), currency: .rub
                )
                
                let tokenizationSettings = TokenizationSettings(paymentMethodTypes: .bankCard)
                
                let paymentParameters = TokenizationModuleInputData(
                    clientApplicationKey: options.clientApplicationKey,
                    shopName: "MedremindR",
                    shopId: options.shopId,
                    purchaseDescription: "Премиальный план",
                    amount: amount,
                    tokenizationSettings: tokenizationSettings,
                    savePaymentMethod: .on
                )
                
                let viewController = TokenizationAssembly.makeModule(
                    inputData: .tokenization(paymentParameters), moduleOutput: self)
                
                self.tokenizationVC = viewController
                currentVC.present(viewController, animated: true)
            }
        
        }
        
    }
    
}
